### ISSUE TITLE
Python and Numba Have Different Behaviors with Jitclass and Deferred Type 

### ISSUE THREAD

**Issue by seanlaw** [numba/numba#3728]
I have the following Python code that I am porting over to Numba:

    class NODE(object):
        def __init__(self):
            self.left = self
            self.right = self
            print(self)
            print(self.left)
            print(self.right)

    if __name__ == "__main__":
        n = NODE()
        print(n)
        print(n.left)
        print(n.right)

For all six print statements, I get the same line printed six times which is expected:

    <__main__.NODE object at 0x1040aa668>
    <__main__.NODE object at 0x1040aa668>
    <__main__.NODE object at 0x1040aa668>
    <__main__.NODE object at 0x1040aa668>
    <__main__.NODE object at 0x1040aa668>
    <__main__.NODE object at 0x1040aa668>

However, porting this over to Numba:

    from numba import deferred_type, jitclass

    node_type = deferred_type()
    node_spec = [("left", node_type),
                 ("right", node_type),
                ]

    @jitclass(node_spec)
    class NODE(object):
        def __init__(self):
            self.left = self
            self.right = self
            print(self)
            print(self.left)
            print(self.right)

    node_type.define(NODE.class_type.instance_type)

    if __name__ == "__main__":
        n = NODE()
        print(n)
        print(n.left)
        print(n.right)

I get the following print statements:

    <numba.jitclass.boxing.NODE object at 0x1101df0f0>
    <numba.jitclass.boxing.NODE object at 0x1101df0d0>
    <numba.jitclass.boxing.NODE object at 0x1101df0f0>
    <numba.jitclass.boxing.NODE object at 0x1101df0d0>
    <numba.jitclass.boxing.NODE object at 0x1101df0f0>
    <numba.jitclass.boxing.NODE object at 0x1101df0f0>

Notice that :

1. They aren't all the same (I expect them to be identical just like the Python only example)
2. `n` does equal `self` and `n.left` does not equal `self.left`
3. Only `n.right` equals `self.right` but they should both be equal to the others

**Comment by alendit** [comment]
When I execute your code, I get 

```
<numba.jitclass.boxing.NODE object at 0x7fb10a17ea30>
<numba.jitclass.boxing.NODE object at 0x7fb10a17ea30>
<numba.jitclass.boxing.NODE object at 0x7fb10a17ea30>
<numba.jitclass.boxing.NODE object at 0x7fb10a17ea30>
<numba.jitclass.boxing.NODE object at 0x7fb1187d0590>
<numba.jitclass.boxing.NODE object at 0x7fb1187d0590>
```

I.e. lines 1-4 are the same, 5-6 differ from 1-4. They are equal, but I think it's spurious. 

The issue here is boxing of the return values of numba functions. To demonstrate the issue even clearer we could write

```
@numba.jitclass([])
class Test:
    def __init__(self):
        pass
   def get_self(self):
       return self
```

Each time you call `get_self` on an instance you would get a different python object, which references the same underlying data. So if your class were to have data fields, python instances would share them.

Of course the lack of identity on returned objects can be considered a bug. We could cache class boxes in a weak value dictionary and check if there is already a box for a class data pointer before creating a new one. @sklam how does it sound?

**Comment by seanlaw** [comment]
For my case, I need to be able to compare two of the objects and see if they are the same. Basically, I need a way to have all of the six values to return True when I compare them. 

Basically, I have a circular doubly linked-list and I'm adding nodes. I start with one node, traverse toward the right until I end up back at my starting node and then stop. First, I store the reference to the starting node and then I keep traversing to the right until the reference to the current node is identical to the starting reference node.

**Comment by alendit** [comment]
The code run in the numba context sees the actual data pointers, so it would know about the objects being equal. Unfortunately, `operator.eq` isn't defined for `jitclass`. You can patch in `eq` support yourself, though:

```
from numba.extending import type_callable, lower_builtin
from numba import cgutils
import operator

@type_callable(operator.eq)
def eq(context):
    """Type the eq operator for jitclasses"""
    def typer(arg1, arg2):
        if isinstance(arg1, types.misc.ClassInstanceType) and isinstance(arg2, types.misc.ClassInstanceType):
            return types.bool_
    return typer

@lower_builtin(operator.eq, types.misc.ClassInstanceType, types.misc.ClassInstanceType)
def lower_node_eq(context, builder, sig, args):
    """Implement comparison operation for jitclasses by comparing their data pointers"""
    retty = sig.return_type
    # see numba.jitclass.base.InstanceDataModel
    get_dataptr = lambda val: builder.extract_value(val, 1)
    ptr1 = get_dataptr(args[0])
    ptr2 = get_dataptr(args[1])
    return builder.icmp_unsigned('==', ptr1, ptr2)
```

Then you could do the following:

```
from numba import deferred_type, jitclass

node_type = deferred_type()
node_spec = [("left", node_type),
             ("right", node_type),
            ]

@jitclass(node_spec)
class Node(object):
    def __init__(self):
        self.left = self
        self.right = self
        
    def get_self(self):
        return self
    
    def equals(self, other):
        return other == self

node_type.define(NODE.class_type.instance_type)

n1 = Node()
n2 = Node()
n1.equals(n1.left)
n1.equals(n1.right)
n1.left.equals(n1.right)
n1.equals(n2)
```

```
True
True
True
False
```

Note, that it would be tempting to define a `__eq__` method on the `Node` class, but dunders aren't supported properly right now (see #1606 and follow #3730).

**Comment by sklam** [comment]
tl;dr summary:
* This is a known behavior of jitclass.
* Jitclass objects in the Python interpreter are wrappers around the underlying numba objects, which have the same data-pointer in the code provided.  
* It's unrelated to `DeferredType`.

details:

A jitclass object in the Python interpreter is a wrapper around the underlying numba object.  The wrapper is being created as part of _boxing_, which is used to wrap a numba object for use as a regular Python object.  Boxing happens to values being returned from a numba function and being used as arguments to `print()`.  (The latter is needed to emulate CPython behavior without re-implementation of the print logic.)  

It's possible to check the pointer of actual numba jitclass object by:

```python
from numba.jitclass import _box
print(_box.box_get_dataptr(n))
```

You will find that `n`, `n.left` and `n.right` have the same data pointer, thus they are identical.


As for:

> They aren't all the same (I expect them to be identical just like the Python only example)
> n does equal self and n.left does not equal self.left
> Only n.right equals self.right but they should both be equal to the others

It's accidental that some Python objects seem to have the same address sometimes.  The CPython allocator tends to reuse available slots.  When an object die, the slot gets reused immediately for the next new object.  Thus, you can have situations like below:

```python
>>> print(id(object()), id(object()))
4408856736 4408856736
>>> print(id(object()), id(object()))
4408856736 4408856736
````
The pointer of the objects are all identical, but they were 4 unique objects.

**Comment by moble** [comment]
> Note that while it would be tempting to define an `__eq__` method on the `Node` class, the dunders aren't supported properly right now (see #1606 and follow #3730).

@alendit Are you suggesting that #3730 will have some impact on dunders other than just `__getitem__` and `__setitem__`?  Is that PR a template that would work for the other dunders?  In particular, any way of using  `__add__`, `__radd__`, `__mul__`, etc., would be completely game-changing for many of us — even if it requires painful low-level programming.

**Comment by alendit** [comment]
Hi Mike,

#3730 only adds get and setitems, but the approach for other dunders is
pretty similar. Now that it's merged, I'll look into adding more of them.

**Comment by moble** [comment]
Thanks @alendit.  I'm *very* excited about this!

**Comment by stuartarchibald** [comment]
Closing this issue as it seems to be resolved. If this is not the case please re-open with a comment about any item that appears to be unresolved. Many thanks.
